sdk/search/Azure.Search.Documents/src/Batching/SearchIndexingBufferedSender{T}.cs

```diff
@@ -0,0 +1,625 @@
+        /// </summary>
+        /// <param name="async">Whether to call this sync or async.</param>
+        /// <returns>A Task that will wait until we're disposed.</returns>
+        internal async Task DisposeAsync(bool async)
+        {
+            if (Interlocked.CompareExchange(ref _disposed, 1, 0) == 0)
+            {
+                await _publisher.DisposeAsync(async).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Ensure the sender was properly disposed.
+        /// </summary>
+        ~SearchIndexingBufferedSender()
+        {
+            if (_publisher?.IndexingActionsCount > 0)
+            {
+                try
+                {
+                    #pragma warning disable CA1065 // Do not raise exceptions in unexpected locations
+                    throw new InvalidOperationException(
+                        $"{nameof(SearchIndexingBufferedSender<T>)} has {_publisher.IndexingActionsCount} unsent indexing actions.");
+                    #pragma warning restore CA1065 // Do not raise exceptions in unexpected locations
+                }
+                catch (InvalidOperationException)
+                {
+                }
+            }
+        }
+
+        /// <summary>
+        /// Ensure nobody's trying to use this after it's been disposed.
+        /// </summary>
+        private void EnsureNotDisposed()
+        {
+            if (_disposed > 0)
+            {
+                throw new ObjectDisposedException(nameof(SearchIndexingBufferedSender<T>));
+            }
+        }
+        #endregion Dispose
+
+        #region Get Key Field
+        /// <summary>
+        /// Ensure we have a valid KeyFieldAccessor (and probe the type or
+        /// service to find one).
+        /// </summary>
+        /// <param name="async">Whether to run sync or async.</param>
+        /// <param name="cancellationToken">The Cancellation token.</param>
+        /// <returns>A task that will complete after we've checked.</returns>
+        private async Task EnsureKeyFieldAccessorAsync(bool async, CancellationToken cancellationToken)
+        {
+            // Skip initialization if we already have one
+            if (KeyFieldAccessor != null) { return; }
+
+            // If we have multiple threads attempting to verify we have a key
+            // field we want to minimize possible requests to the service for
+            // its index.  We'll only assign the task if it's not null and
+            // anyone who gets year later can wait on it with us.
+            _getKeyFieldAccessorTask ??= GetKeyFieldAccessorAsync(async, cancellationToken);
+            if (async)
+            {
+                await _getKeyFieldAccessorTask.ConfigureAwait(false);
+            }
+            else
+            {
+                #pragma warning disable AZC0104 // Use EnsureCompleted() directly on asynchronous method return value.
+                _getKeyFieldAccessorTask.EnsureCompleted();
```

Comment:
What if the first call to `EnsureKeyFieldAccessorAsync` was async and the task is in progress?

Reply:
Great point.  I'd _really_ hope nobody ever tried that...  but changed to do sync-over-async in that case.